capreolus/extractor/__init__.py

```diff
@@ -1,20 +1,170 @@
+    embed_pathes = {
+        "glove6b": "glove/light/glove.6B.300d",
+        "glove6b.50d": "glove/light/glove.6B.50d",
+        "w2vnews": "word2vec/light/GoogleNews-vectors-negative300",
+        "fasttext": "fasttext/light/wiki-news-300d-1M-subword",
+    }
+
     @staticmethod
     def config():
-        keepstops = False
+        embeddings = "glove6b"
+        zerounk = False
+        calcidf = True
+        maxqlen = 4
+        maxdoclen = 7
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.itos = {self.pad: self.pad_tok}
+        self.stoi = {self.pad_tok: self.pad}
+        self.qid2toks = defaultdict(list)
+        self.docid2toks = defaultdict(list)
+        self.idf = defaultdict(lambda: 0)
+        self.embeddings = None
+        # self.cache = self.load_cache()    # TODO
+
+    def _get_pretrained_emb(self):
+        return Magnitude(
+            MagnitudeUtils.download_model(self.embed_pathes[self.cfg["embeddings"]], download_dir=self.get_cache_path())
+        )
+
+    def _build_vocab(self, qids, docids, topics):
+        tokenize = self["tokenizer"].tokenize
+        self.qid2toks = {qid: tokenize(topics.get(qid, "")) for qid in qids}
+        self.docid2toks = {docid: tokenize(self["index"].get_doc(docid)) for docid in docids}
+        self._extend_stoi(self.qid2toks.values(), calc_idf=self.cfg["calcidf"])
+        self._extend_stoi(self.docid2toks.values(), calc_idf=self.cfg["calcidf"])
+        self.itos = {i: s for s, i in self.stoi.items()}
+        logger.info(f"vocabulary constructued, with {len(self.itos)} terms in total")
+
+    def _get_idf(self, toks):
+        return [self.idf.get(tok, 0) for tok in toks]
+
+    def _build_embedding_matrix(self):
+        assert len(self.stoi) > 1  # needs more vocab than self.pad_tok
+
+        magnitude_emb = self._get_pretrained_emb()
+        emb_dim = magnitude_emb.dim
+        embed_vocab = set(term for term, _ in magnitude_emb)
+        embed_matrix = np.zeros((len(self.stoi), emb_dim), dtype=np.float32)
+
+        n_missed = 0
+        for term, idx in self.stoi.items():
+            if term in embed_vocab:
+                embed_matrix[idx] = magnitude_emb.query(term)
+            elif term == self.pad_tok:
+                embed_matrix[idx] = np.zeros(emb_dim)
+            else:
+                n_missed += 1
+                embed_matrix[idx] = np.zeros(emb_dim) if self.cfg["zerounk"] else np.random.normal(scale=0.5, size=emb_dim)
+
+        logger.info(f"Embedding matrix {self.cfg['embeddings']} constructued, with shape {embed_matrix.shape}")
+        if n_missed > 0:
+            logger.warning(f"{n_missed}/{len(self.stoi)} (%.3f) term missed" % (n_missed / len(self.stoi)))
+
+        self.embeddings = embed_matrix
+
+    def exist(self):
+        return isinstance(self.embeddings, np.ndarray) and 0 < len(self.stoi) == self.embeddings.shape[0]
+
+    def create(self, train_pairs, pred_pairs, topics):
```

Comment:
Regarding point-1: I think the way this works now is reasonable (and better than passing benchmark), because the extractor has no way of knowing how train_pairs and pred_pairs will be built